rfcs/20190815-tfx-notebook.md

```diff
@@ -0,0 +1,563 @@
+*   Simple Python API per TFX component
+    *   ExampleGen, StatsGen, SchemaGen, Transform, Trainer, Evaluator
+    *   100% TFX compatible for automatic notebook → pipeline export
+*   Analyze artifacts natively in Notebook/Colab
+    *   Built-in TensorBoard, Facets, TFMA visualizations
+    *   Dataset, stats, eval metrics available in notebook for custom analysis
+*   Zero-setup, interactive onboarding tool for new TFX users on
+    [tensorflow.org](http://tensorflow.org)
+
+## Motivation
+
+The benefits of using a notebook include rapidly editing and running code,
+immediately seeing the execution and outputs of commands, and running quick
+one-off analyses in Python. It’s a simple, no-mental-overhead REPL environment
+for iterating on ideas.
+
+By combining the notebook experience + TFX components, users can easily run
+
+*   ExampleGen to generate the initial dataset used for training
+*   StatsGen to generate and visualize a statistical report of your data
+*   SchemaGen to generate a schema of your data (required input of Transform)
+*   Transform to write feature engineering strategies
+*   Trainer that wraps standard TF.Estimator or Keras code
+*   Evaluator to generate, slice, and visualize evaluation metrics
+*   Custom analyses on the output of any of these components with standard
+    Python
+
+To close the loop, the notebook will be automatically exported as a pipeline
+configuration that users can directly deploy as a scalable TFX pipeline. There
+is no additional modification required.
+
+## Target Users
+
+We target users who want to manually iterate on their models & components, and
+prefer a notebook environment for the benefits outlined above. This is a wide
+range of potential users, and from our user research, spans software engineers
+and data scientists within and outside of Google.
+
+## Design Proposal
+
+This proposal proposes a set of primitives that match concepts in the current
+TFX SDK.
+
+### Component definition; inputs and outputs
+
+#### Proposal: components should take inputs, produce outputs (instead of taking predefined upstream components)
+
+This proposal proposes a set of primitives that match concepts in the current
+TFX SDK. We propose to follow the current TFX style of having components
+explicitly take input channels (i.e. streams of artifacts of a specific type)
+and produce output channels (of another specific type). This could look like
+this:
+
+```
+# Here, with an input_base as an execution parameter with a given
+# file path.
+example_gen = CsvExampleGen(input_base=examples)
+
+# Next, we use the 'examples' named output of ExampleGen as the
+# input to StatisticsGen.
+statistics_gen = StatisticsGen(input_data=example_gen.outputs['examples'])
+
+# We then similarly use the statsgen output in SchemaGen.
+infer_schema = SchemaGen(stats=statistics_gen.outputs['output'])
+
+# Next, we do example validation.
+validate_stats = ExampleValidator(
+    stats=statistics_gen.outputs['output'],
```

Comment:
Needing to call .outputs['output'] seems unfortunate-- is there a convenience method or structuring that will make this less redundant for the default use-case?